### PR TITLE
GCI: Added Tensorflow docs

### PR DESCRIPTION
**Pull request by seventeen-iterator**
Added another documentation about usage of exported models in tensorflow.

### CODE REVIEW

**Comment by thatbrguy** [comment]
Hello @seventeen-iterator good work so far. However, most of the code had been covered in the previous PR right, so we would want something more novel.

Here's what we can try. Write code for sample usage for your own neural network model. You can create a simple CNN for instance using instructions in the same [blog post](https://towardsdatascience.com/cifar-10-image-classification-in-tensorflow-5b501f7dc77c). 

Do the above in a separate python file. Update TODO in your current python file saying "Once graph export it is fixed, usage functionality from the new file can be inserted here".

**Comment by seventeen-iterator** [comment]
Hello, I added CNN that trained on the fashion MNIST dataset. I'm not sure how good is the model I created, but more than enough for a sample.

File: example/tensorflow/code_template/tensorflow_fashion_mnist.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
+from __future__ import print_function
+
+import tensorflow as tf
+from tensorflow.keras.datasets import fashion_mnist
+
+(train_x, train_y), (test_x, test_y) = fashion_mnist.load_data()
```
Comment (thatbrguy) [inline example/tensorflow/code_template/tensorflow_fashion_mnist.py]: Move lines 6,8 and 10 to after function definition.

File: example/tensorflow/code_template/tensorflow_fashion_mnist.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+    dense = tf.layers.dense(flat, 128, activation=tf.nn.tanh)
+    dense = tf.nn.dropout(dense, keep_prob)
+    dense = tf.layers.batch_normalization(dense)
+
+    out = tf.layers.dense(dense, 10, activation=None)
+    return out
+
+
+def train_neural_network(session, optimizer, keep_probability, feature_batch, label_batch):
```
Comment (thatbrguy) [inline example/tensorflow/code_template/tensorflow_fashion_mnist.py]: You can rename `session` to `sess` for consistency with rest of the code. Add one liner comment on the purpose of the function.

File: example/tensorflow/code_template/tensorflow_fashion_mnist.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    session.run(optimizer,
+                feed_dict={
+                    x: feature_batch,
+                    y: label_batch,
+                    keep_prob: keep_probability
+                })
+
+
+def print_stats(sess, feature_batch, label_batch,
```
Comment (thatbrguy) [inline example/tensorflow/code_template/tensorflow_fashion_mnist.py]: Add one liner comment on the purpose of the function.

File: example/tensorflow/code_template/tensorflow_fashion_mnist.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+epochs = 10
+batch_size = 128
+n_batches = train_x.shape[0] // batch_size + 1
+keep_probability = 0.7
+learning_rate = 0.001
+
+x = tf.placeholder(tf.float32, shape=(None, 28, 28), name='input_x')
+depth = 10
+y = tf.placeholder(tf.uint8, shape=(None,), name='output_y')
+check_y = tf.one_hot(y, depth)
```
Comment (thatbrguy) [inline example/tensorflow/code_template/tensorflow_fashion_mnist.py]: Can use a better variable name other than `check_y`.

File: example/tensorflow/code_template/tensorflow_fashion_mnist.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+        for batch_i in range(n_batches):
+            batch_start = batch_i*batch_size
+            batch_end = batch_start+batch_size
+            batch_features = train_x[batch_start:batch_end]
+            batch_labels = train_y[batch_start:batch_end]
+
+            train_neural_network(sess, optimizer, keep_probability, batch_features, batch_labels)
+
+            print('Epoch {:>2}, Fashion MNIST Batch {}:  '.format(epoch + 1, batch_i + 1), end='')
+            print_stats(sess, batch_features, batch_labels, test_x, test_y, cost, accuracy)
```
Comment (thatbrguy) [inline example/tensorflow/code_template/tensorflow_fashion_mnist.py]: Two things:
1. Instead of calling it `batch_features` and `batch_labels`, we can call it `train_features` and `train_labels`. 
2. Accuracy and loss must be averaged over the entire epoch (outer loop) and then only printed. Please try to make that change.
3. In your code, validation loss and accuracy is calculated over the entire test dataset (`test_x` and `test_y`). You may have to consider evaluating the accuracy in batches just like how you do for the training set, as datasets can typically be large. What you can do is, have a variable named `n_batches_validation` and calculate validation loss and accuracy using that loop, after `n_batches` of training is done. Consider making this change.

**Comment by thatbrguy** [comment]
Hello @seventeen-iterator I have made some suggestions on your submission. Check it out.

File: example/tensorflow/code_template/tensorflow_fashion_mnist.py (hunk 5 above)
Comment (seventeen-iterator) [reply example/tensorflow/code_template/tensorflow_fashion_mnist.py]: But in this code the loss is calculated over training batch, not validation set. Maybe I need to consider separating print_stats function, since the loss and accuracy shall be calculated in the batch loop, but stats must be printed only for every epoch.

Comment (thatbrguy) [reply example/tensorflow/code_template/tensorflow_fashion_mnist.py]: - My bad, I meant to say accuracy over validation set, not loss. However, it's okay to calculate both loss and accuracy for training and validation. 
- Yes, try modifying or removing the function `print_stats` and calculate accuracy and loss for both train and validation.

**Comment by seventeen-iterator** [comment]
Also added some progress bars to make it look nearly like keras

File: example/tensorflow/code_template/tensorflow_fashion_mnist.py [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+        'train loss: {:4}, train acc: {:4},'
+        'valid loss: {:4}, valid acc: {:4}'
+    ).format(
+        epoch,
+        train_losses.mean(), train_accs.mean(),
+        valid_losses.mean(), valid_accs.mean()
+    ))
+
+
+(train_x, train_y), (valid_x, valid_y) = fashion_mnist.load_data()
```
Comment (Ram81) [inline example/tensorflow/code_template/tensorflow_fashion_mnist.py]: @seventeen-iterator can you add the code below in methods too, so that it is easy to understand.

Comment (seventeen-iterator) [reply example/tensorflow/code_template/tensorflow_fashion_mnist.py]: Do you mean, all code after highlighted lines?

Comment (seventeen-iterator) [reply example/tensorflow/code_template/tensorflow_fashion_mnist.py]: Oh, you talk about print stats function. For some reason I could see this only in email.

**Comment by seventeen-iterator** [comment]
Is this easy enough to understand now?

**Comment by thatbrguy** [comment]
Hello @seventeen-iterator, code looks good. CLI logger also looks good. I'm attaching a screenshot here for other mentors reference:

tf

Two things:

- I was not able to execute `from keras.datasets import fashion_mnist` for python2. Works for python3 though. Did you encounter such an issue?
- Mention your fashion mnist code in `tensorflow_pbtext_usage.md` along with a description of what it does. Don't forget to mention users to `pip install tqdm`

File: tutorials/tensorflow_pbtext_usage.md [hunk 7]
```diff
@@ -0,0 +1,31 @@
+# How to use pbtxt to import tensorflow model
+
+1. Select the 3nd button in the Actions section:
```
Comment (Ram81) [inline tutorials/tensorflow_pbtext_usage.md]: Make this `3rd` instead of `3nd`

File: tutorials/tensorflow_pbtext_usage.md [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+
+    ![](select_export_new.png)
+2. In appeared dropdown list select Tensorflow. This should download a `.pbtxt` file
+
+    ![](select_tensorflow_red.png)
+3. Rename the file to `model.pbtxt`
+4. In the folder with model file this code:
```
Comment (Ram81) [inline tutorials/tensorflow_pbtext_usage.md]: Can't interpret the meaning of this line, can you rephrase this.

**Comment by seventeen-iterator** [comment]
@thatbrguy Hi! I wasn't launching it first, because I didn't have python2. Also tensorflow doesn't exist for Windows + Python2. But I managed to launch `tensorflow_fashion_mnist.py` on the virtual machine, and no errors caught. (I installed anaconda, tensorflow and tqdm first)
The rest I'll satisfy later.

**Comment by thatbrguy** [comment]
@seventeen-iterator Oh you're using windows. Alright then. Address the other changes before your time runs out! (I believe you have 5 hours or so left).

**Comment by seventeen-iterator** [comment]
Both lines executed without errors in python2.
